src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,169 @@
+        for cluster in self.catalog().clusters() {
+            if let ClusterVariant::Managed(ref config) = cluster.config.variant {
+                match config.schedule {
+                    ClusterScheduleOptionValue::Manual => {
+                        // nothing to do, user manages it manually
+                    }
+                    ClusterScheduleOptionValue::Refresh => {
+                        let refresh_mvs = cluster
+                            .bound_objects()
+                            .iter()
+                            .filter_map(|id| {
+                                if let CatalogItem::MaterializedView(mv) =
+                                    self.catalog().get_entry(id).item()
+                                {
+                                    if mv.refresh_schedule.is_some() {
+                                        Some(id)
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect_vec();
+
+                        // Check whether there is at least one REFRESH MV needing a refresh.
+                        let local_read_ts = self.get_local_read_ts().await;
+                        decisions.push((
+                            cluster.id,
+                            refresh_mvs.into_iter().any(|mv| {
+                                let mv_frontier = &self
+                                    .controller
+                                    .storage
+                                    .collection(*mv)
+                                    .expect("the storage controller should know about MVs that exist in the catalog")
+                                    .write_frontier;
+                                mv_frontier.less_than(&local_read_ts)
+                            }),
+                        ));
+                    }
+                }
+            }
+        }
+        decisions
+    }
+
+    #[mz_ore::instrument(level = "debug")]
+    pub(crate) async fn handle_scheduling_decisions(
+        &mut self,
+        decisions: Vec<(&'static str, Vec<(ClusterId, bool)>)>,
+    ) {
+        for (policy_name, decisions) in decisions.iter() {
+            for (cluster_id, decision) in decisions {
+                self.scheduling_decisions
+                    .entry(*cluster_id)
+                    .or_insert_with(Default::default)
+                    .insert(policy_name, *decision);
+            }
+        }
+
+        // Clean up those clusters from `scheduling_decisions` that
+        // - have been dropped, or
+        // - were switched to unmanaged, or
+        // - were switched to `SCHEDULE = MANUAL`.
+        for cluster_id in self.scheduling_decisions.keys().cloned().collect_vec() {
```

Comment:
Is there a reason you're collecting this iter? `.cloned()` should be enough to clone the keys.

Reply:
I collected it because the loop is also modifying `scheduling_decisions`. With just calling `.cloned()`, the iterator would still retain a reference to `scheduling_decisions`.